core/Http/RequestSanitizer.php

```diff
@@ -0,0 +1,233 @@
+
+/**
+ * Service that sanitizes query parameters for requests.
+ *
+ * Eventually once consistent out-only sanitization is implemented, this class should be
+ * removed. In the mean time, since this class is accessed through DI, it can be used
+ * to implement consistent sanitization, piece by piece.
+ */
+class RequestSanitizer
+{
+    // Flag used with htmlspecialchar. See php.net/htmlspecialchars.
+    const HTML_ENCODING_QUOTE_STYLE = ENT_QUOTES;
+
+    /**
+     * @var array[]
+     */
+    private $requestParameterSanitizeBlacklist = array();
+
+    public function __construct(array $requestParameterSanitizeBlacklist)
+    {
+        foreach ($requestParameterSanitizeBlacklist as $entry) {
+            list($apiMethod, $params) = $entry;
+
+            if (isset($this->requestParameterSanitizeBlacklist[$apiMethod])) {
+                $this->requestParameterSanitizeBlacklist[$apiMethod] =
+                    array_merge($this->requestParameterSanitizeBlacklist[$apiMethod], $params);
+            } else {
+                $this->requestParameterSanitizeBlacklist[$apiMethod] = $params;
+            }
+        }
+    }
+
+    /**
+     * Sanitizes a single value.
+     *
+     * @param mixed $value If numeric, the value is not changed. If a string, line breaks are removed and HTML entities
+     *                     are escaped.
+     * @param bool $alreadyStripslashed Implementation detail, ignore.
+     * @return string
+     * @throws \Exception If the value is null or a boolean.
+     */
+    public function sanitizeValue($value, $alreadyStripslashed = false)
+    {
+        if (is_numeric($value)) {
+            return $value;
+        } else if (is_string($value)) {
+            $value = $this->sanitizeLineBreaks($value);
+            $value = $this->sanitizeString($value);
+
+            if (!$alreadyStripslashed) { // a JSON array was already stripslashed, don't do it again for each value
+                $value = $this->undoMagicQuotes($value);
+            }
+        } else if (!is_null($value)
+            && !is_bool($value)
+        ) {
+            throw new \Exception("The value to escape has not a supported type. Value = " . var_export($value, true));
+        }
+        return $value;
+    }
+
+    /**
+     * Unsanitizes a single value. Replaces escaped HTML entities with the actual characters.
+     *
+     * @param string $value
+     * @return string
+     */
+    public function unsanitizeValue($value)
+    {
+        return htmlspecialchars_decode($value, self::HTML_ENCODING_QUOTE_STYLE);
+    }
+
+    /**
+     * Sanitize an array of values recursively.
+     *
+     * @param mixed $value If an array, it is sanitized recursively. Keys are sanitized as well as values. If not,
+     *                     it is sanitized normally.
+     * @param bool $alreadyStripslashed Implementation detail, ignore.
+     * @return mixed
+     */
+    public function sanitizeValues($value, $alreadyStripslashed = false)
```

Comment:
This could be merged with `sanitize()` because it already accepts `mixed` so there's not really a reason to have 2 methods.